VMware Module to perform operations on VNIC

mohdumar321 opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->

 - New Module Pull Request
- created a new module for performing operations on vmware vnics.
- tested 


##### COMPONENT NAME
<!--- vmware_vnic.py-->
vmware_vnic.py
##### ANSIBLE VERSION
<!--- 2.x-->
```
2.x
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

mohdumar321 committed:
commit

mohdumar321 committed:
deletetesting

mohdumar321 committed:
commit

mohdumar321 committed:
commit

mohdumar321 committed:
commit

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vnic.py:
```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
```
Add license boilerplate

```
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vnic.py:
```diff
@@ -0,0 +1,137 @@
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': ''}
```
`'supported_by': 'community'}`

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vnic.py:
```diff
@@ -0,0 +1,137 @@
+
+DOCUMENTATION = '''
+---
+module: vmware_vnic
```
I would recommend to rename this to `vmware_guest_nic_manager` or `vmware_guest_nic` or something, this will match will existing naming scheme.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vnic.py:
```diff
@@ -0,0 +1,137 @@
+DOCUMENTATION = '''
+---
+module: vmware_vnic
+short_description: Manage Operations on vmware vnics
```
Please add description and related documentation as this will be used by both docs.ansible.com and ansible-doc command.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vnic.py:
```diff
@@ -0,0 +1,137 @@
+
+def get_args():
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(dict(hostname=dict(type='str', required=True),
```
hostname, username, password and port are already part of `vmware_argument_spec()`

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vnic.py:
```diff
@@ -0,0 +1,137 @@
+                              username=dict(type='str', required=True),
+                              password=dict(type='str', required=True),
+                              port=dict(type='int', default='443'),
+                              name=dict(type='str', required=True),
```
virtual machine name and uuid are mutual exclusive parameters and cannot be required.

Akasurde replied:
String is default argument type so not necessary.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vnic.py:
```diff
@@ -0,0 +1,137 @@
+                              port=dict(type='int', default='443'),
+                              name=dict(type='str', required=True),
+                              uuid=dict(type='str', required=True),
+                              nic_state=dict(type='str'),
```
I would rename this simply as `state`

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vnic.py:
```diff
@@ -0,0 +1,137 @@
+    return module
+
+
+def get_obj(content, vim_type, name):
```
Use existing method from `module_utils/vmware.py`

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vnic.py:
```diff
@@ -0,0 +1,137 @@
+    args = get_args()
+
+    # connect to vc
+    si = SmartConnect(
```
use existing method `connect_to_api` from module_utils/vmware.py

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vnic.py:
```diff
@@ -0,0 +1,137 @@
+
+    content = si.RetrieveContent()
+    print ('Searching for VM {}').format(args.params['name'])
+    vm_obj = get_obj(content, [vim.VirtualMachine], args.params['name'])
```
use `PyVmomi` helper class from `module_utils/vmware.py`

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vnic.py:
```diff
@@ -0,0 +1,137 @@
+
+    if vm_obj:
+        update_virtual_nic_state(si, vm_obj, args.params['nic_number'], args.params['nic_state'])
+        print ('VM NIC {} successfully' \
```
Don't use print statements in module.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vnic.py:
```diff
@@ -0,0 +1,137 @@
+    virtual_nic_spec.operation = \
+        vim.vm.device.VirtualDeviceSpec.Operation.remove \
+        if new_nic_state == 'delete' \
+        else vim.vm.device.VirtualDeviceSpec.Operation.edit
```
Can we add new NICs as well ?

mohdumar321 replied:
@Akasurde : Thanks for the valuable suggestions. I will update as suggested.

Currently we can delete , connect and disconnect vnics but later we can add functionality to add vnics as well.

Akasurde commented:
@mohdumar321 Are you working on this ? needs_info

mohdumar321 commented:
@akasurde: I am almost done will commit soon.

mohdumar321 committed:
commit with new changes- - Module name change  -Documentation update - Boilerplate added - Example updated - Removed print statements

Done changes 
- Module Name Change
- Documentation update
- Boilerplate added
- Example updated
- Removed print statements

mohdumar321 commented:
@Akasurde: I have committed code with changes. Unable to identify where it got failed Please review it and if needed please update because I am unable to identify.

mohdumar321 committed:
commit

mohdumar321 committed:
commit

mohdumar321 committed:
commit

mohdumar321 committed:
commit

mohdumar321 committed:
commit

Akasurde commented:
superseded by https://github.com/ansible/ansible/pull/52075

mohdumar321 commented:
@Akasurde : Thanks
